src/widgets/calendar/lv_calendar_lunar.c

```diff
@@ -0,0 +1,243 @@
+    {"元宵节", 1, 15},
+    {"端午节", 5, 5},
+    {"七夕节", 7, 7},
+    {"中秋节", 8, 15},
+    {"重阳节", 9, 9},
+};
+
+static lv_calendar_festival solar_festival[] = {
+    {"元旦", 1, 1},
+    {"情人节", 2, 14},
+    {"妇女节", 3, 8},
+    {"植树节", 3, 12},
+    {"消费节", 3, 15},
+    {"愚人节", 4, 1},
+    {"劳动节", 5, 1},
+    {"青年节", 5, 4},
+    {"儿童节", 6, 1},
+    {"建党节", 7, 1},
+    {"建军节", 8, 1},
+    {"教师节", 9, 10},
+    {"国庆节", 10, 1},
+    {"万圣节", 10, 31},
+    {"平安夜", 12, 24},
+    {"圣诞节", 12, 25},
+};
+
+static char buff[15];
+
+/**********************
+ *      MACROS
+ **********************/
+
+/**********************
+ *   GLOBAL FUNCTIONS
+ **********************/
+
+void lv_calendar_set_lunar_mode(lv_obj_t * obj, bool en)
+{
+    lv_calendar_t * calendar = (lv_calendar_t *)obj;
+    calendar->use_lunar_calendar = en;
+    lv_calendar_set_showed_date(obj, calendar->today.year, calendar->today.month);
+}
+
+char * lv_calendar_get_current_day_name(lv_calendar_date_t * solar)
+{
+    uint16_t i, len;
+    lv_calendar_lunar lunar_date;
+    lunar_date = Solar_to_Lunar(solar);
+    if(solar->year > 2099 || solar->year < 1901)
+        return NULL;
+    len = sizeof(lunar_festival) / sizeof(lv_calendar_festival);
+    for(i = 0; i < len; i++) {
+        if((lunar_date.today.month == lunar_festival[i].month) &&
+           (lunar_date.today.day == lunar_festival[i].day) &&
+           lunar_date.leep_month == false)
+            return (char *)lunar_festival[i].festival_name;
+    }
+    len = sizeof(solar_festival) / sizeof(lv_calendar_festival);
+    for(i = 0; i < len; i++) {
+        if((solar->month == solar_festival[i].month) &&
+           (solar->day == solar_festival[i].day))
+            return (char *)solar_festival[i].festival_name;
+    }
+    if(lunar_date.today.day == 1) {
+        if(lunar_date.leep_month == false)
+            return (char *)LunarMonth_name[lunar_date.today.month - 1];
+        else {
+            lv_snprintf(buff,
+                        sizeof(buff),
+                        "闰%s",
+                        LunarMonth_name[lunar_date.today.month - 1]);
+            return buff;
+        }
+    }
+    return (char *)LunarDay_name[lunar_date.today.day - 1];
+}
+
+/**********************
+ *  STATIC FUNCTIONS
+ **********************/
+
+static lv_calendar_lunar Solar_to_Lunar(lv_calendar_date_t * solar)
+{
+    uint16_t year = solar->year;
+    uint8_t month = solar->month;
+    uint8_t day = solar->day;
+
+    /*Record the number of days between the Spring Festival
+    and the New Year's Day of that year.*/
+    uint16_t bySpring;
+
+    /*Record the number of days from the solar calendar
+    to the New Year's Day of that year.*/
+    uint16_t bySolar;
+
+    /*Record the number of days in that month*/
+    uint8_t daysPerMonth;
+
+    /*Record from which month the calculation starts.*/
+    uint8_t index;
+
+    bool leep_month;
+    lv_calendar_lunar lunar;
+
+    if(((calendar_lunar_table[year - 1901] & 0x0060) >> 5) == 1)
```

Comment:
Please add a guard somewhere before this array access to ensure `year` is in the range [1901..2099] inclusive.